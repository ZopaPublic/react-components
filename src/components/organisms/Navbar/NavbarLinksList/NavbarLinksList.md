### Summary

[`<Navbar.LinksList />`](#/Components/Organisms/Navbar/NavbarLinksList) is responsible for rendering navbar links. By default it will render [`<Navbar.Dropdown />`](#/Components/Organisms/Navbar/NavbarDropdown) for `links` with `children` and [`<Navbar.Link />`](#/Components/Organisms/Navbar/NavbarLink) for `links` within a dropdown (if using `<Navbar.Dropdown />`) or `links` without `children`. These can be changed using the `renderDropdown` and `renderLink` props.

### Examples

- Default

```ts
import { Navbar } from '@zopauk/react-components';

const NAV_ITEMS = [
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

<Navbar.LinksList links={NAV_ITEMS} />;
```

- Custom components

```ts
import { Navbar, Link } from '@zopauk/react-components';

const NAV_ITEMS = [
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

<Navbar.LinksList
  links={NAV_ITEMS}
  renderDropdown={(item, index) => (
    <div>
      Custom Dropdown
      <br /> {item.children.map(link => `${link.label} `)}
    </div>
  )}
  renderLink={(item, props) => (
    <Link href={item.href} target="_blank" {...props}>
      {item.label}
    </Link>
  )}
/>;
```
