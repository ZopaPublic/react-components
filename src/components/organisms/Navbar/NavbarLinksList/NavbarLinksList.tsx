import React from 'react';

import NavbarDropdown, { RenderItemProps } from '../NavbarDropdown/NavbarDropdown';
import { NavigationItem, NavbarLinksListProps } from '../Navbar/Navbar';
import styled from 'styled-components';
import { colors, spacing } from '../../../../constants';
import { minMedia } from '../../../../helpers/responsiveness';

const SingleLink = styled.li`
  display: inline-block;
  border-bottom: 1px solid ${colors.greyLighter};
  margin-bottom: ${spacing[3]};
  padding-bottom: ${spacing[2]};

  ${minMedia.desktop`
    border-bottom: 0;
    margin-bottom: 0;
    padding-bottom: 0;
  `}

  &:last-child {
    border-bottom: 0;
    margin-bottom: 0;
    padding-bottom: 0;
  }
`;

const NavbarLinksList = ({ links, renderLink, setOpen }: NavbarLinksListProps) => {
  return (
    <>
      {links &&
        links.map((item: NavigationItem, index: number) =>
          item.children ? (
            <NavbarDropdown
              key={`dropdown-${index}`}
              id={`navbar-dropdown-${index}`}
              label={item.label}
              items={item.children!}
              data-automation={item['data-automation']}
              renderItem={({ item, getItemProps, close }: RenderItemProps) =>
                renderLink &&
                renderLink(item, index, {
                  ...getItemProps(),
                  isDropdownLink: true,
                  onClick: (e: React.MouseEvent<HTMLAnchorElement>) => {
                    close();
                    setOpen(false);
                    item.onClick && item.onClick(e);
                  },
                })
              }
            />
          ) : (
            <SingleLink key={`link-${index}`}>
              {renderLink &&
                renderLink(item, index, {
                  onClick: (e: React.MouseEvent<HTMLAnchorElement>) => {
                    setOpen(false);
                    item.onClick && item.onClick(e);
                  },
                })}
            </SingleLink>
          ),
        )}
    </>
  );
};

export default NavbarLinksList;
