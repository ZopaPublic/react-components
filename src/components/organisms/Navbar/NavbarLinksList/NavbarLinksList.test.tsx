import React from 'react';
import { render, screen } from '@testing-library/react';
import NavbarLinksList from './NavbarLinksList';

describe('<NavbarLinksList />', () => {
  it('should render component with default props', () => {
    const NAV_ITEMS = [
      {
        label: 'About',
        qadata: 'About.topBar.Menu',
        children: [
          {
            label: 'About Zopa',
            href: '/about',
            qadata: 'About.About.topBar.Menu',
          },
          {
            label: 'Our story',
            href: '/about/our-story',
            qadata: 'Story.About.topBar.Menu',
          },
          {
            label: 'Our board',
            href: '/about/board',
            qadata: 'Our_Board.About.topBar.Menu',
          },
          {
            label: 'Our leadership team',
            href: '/about/leadership',
            qadata: 'Our_Leadership.About.topBar.Menu',
          },
          {
            label: 'Awards',
            href: '/about/awards',
            qadata: 'Awards.About.topBar.Menu',
          },
          {
            label: 'Careers',
            href: '/about/careers',
            qadata: 'Careers.About.topBar.Menu',
          },
          {
            label: 'Press office',
            href: '/about/press',
            qadata: 'Press_office.About.topBar.Menu',
          },
        ],
      },
      {
        label: 'Support',
        href: '/contact',
        qadata: 'Support.topBar.Menu',
      },
    ];
    const container = render(
      <NavbarLinksList
        setOpen={(open) => open}
        links={NAV_ITEMS}
        renderLink={(item) => <a href={item.href}>{item.label}</a>}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render a navlink when links contain children', () => {
    const NAV_ITEMS = [
      {
        label: 'Support',
        href: '/contact',
        qadata: 'Support.topBar.Menu',
      },
    ];

    render(
      <NavbarLinksList
        setOpen={(open) => open}
        links={NAV_ITEMS}
        renderLink={(item) => <a href={item.href}>{item.label}</a>}
      />,
    );
    const menuItem = screen.getByText('Support');
    expect(menuItem).toHaveAttribute('href', '/contact');
  });
});
