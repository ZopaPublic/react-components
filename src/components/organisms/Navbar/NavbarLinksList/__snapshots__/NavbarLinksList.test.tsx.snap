// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavbarLinksList /> should render component with default props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c5 {
  position: relative;
  z-index: 1;
  margin: 12px 0 12px;
  padding: 0;
  list-style-type: none;
  border-bottom: 1px solid #EFEFEF;
}

.c1 {
  background-color: transparent;
  font-size: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  line-height: inherit;
  color: #3B46C4;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-user-select: auto;
  -moz-user-select: auto;
  -ms-user-select: auto;
  user-select: auto;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  border: none;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  color: auto;
}

.c1:hover,
.c1:active {
  color: #21247F;
}

.c1:hover svg path,
.c1:active svg path {
  fill: #21247F;
}

.c1:hover {
  color: auto;
}

.c1:hover svg path {
  fill: auto;
}

.c1:active {
  color: auto;
}

.c1:active svg path {
  fill: auto;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 600;
  color: #3B46C4;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 12px 16px 16px;
}

.c2:focus {
  outline: none;
  box-shadow: 0px 0px 4px 0px #3B46C4;
  border-radius: 4px;
}

.c2:active,
.c2:hover {
  color: #21247F;
  opacity: 0.88;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-right: 8px;
}

.c4 {
  display: none;
}

.c0 {
  position: relative;
  display: inline-block;
}

.c6 {
  display: inline-block;
  border-bottom: 1px solid #EFEFEF;
  margin-bottom: 12px;
  padding-bottom: 8px;
}

.c6:last-child {
  border-bottom: 0;
  margin-bottom: 0;
  padding-bottom: 0;
}

@media (min-width:1280px) {
  .c5 {
    position: absolute;
    left: 50%;
    top: 50px;
    background-color: #FFFFFF;
    box-shadow: 0px 1px 0px 1px #D4D7D9;
    border: 1px solid #EFEFEF;
    border-radius: 12px;
    min-width: 240px;
    margin: 0;
    padding: 8px 0;
    -webkit-transition: opacity 0.3s,-webkit-transform 0.3s,visibility 0.3s 0.3s;
    -webkit-transition: opacity 0.3s,transform 0.3s,visibility 0.3s 0.3s;
    transition: opacity 0.3s,transform 0.3s,visibility 0.3s 0.3s;
    opacity: 0;
    visibility: hidden;
    -webkit-transform: translate(-50%,-10%);
    -ms-transform: translate(-50%,-10%);
    transform: translate(-50%,-10%);
  }

  .c5:before {
    content: '';
    top: 1px;
    left: 50%;
    position: absolute;
    width: 25px;
    height: 25px;
    -webkit-transform: translate(-50%,-50%) rotate(135deg);
    -ms-transform: translate(-50%,-50%) rotate(135deg);
    transform: translate(-50%,-50%) rotate(135deg);
    background-color: inherit;
    box-shadow: -1px 1px 0 0 #EFEFEF;
  }
}

@media (max-width:1279px) {
  .c2 {
    width: 100%;
    color: #2C3236;
    font-size: 14px;
    font-weight: 700;
    text-transform: uppercase;
  }
}

@media (min-width:1280px) {
  .c4 {
    display: -webkit-inline-box;
    display: -webkit-inline-flex;
    display: -ms-inline-flexbox;
    display: inline-flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    font-size: 12px;
    -webkit-transition: -webkit-transform 0.3s;
    -webkit-transition: transform 0.3s;
    transition: transform 0.3s;
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }
}

@media (min-width:1280px) {
  .c6 {
    border-bottom: 0;
    margin-bottom: 0;
    padding-bottom: 0;
  }
}

<body>
    <div>
      <li
        class="c0"
      >
        <a
          aria-expanded="false"
          aria-haspopup="true"
          class="c1 c2"
          color="#3B46C4"
          data-automation="ZA.navbar-item"
          href="#"
          tabindex="0"
        >
          <span
            class="c3"
          >
            About
          </span>
          <span
            class="c4"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              color="#818F9B"
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              height="12px"
              role="img"
              viewBox="0 0 448 512"
              width="12px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </span>
        </a>
        <ul
          aria-label="About"
          class="c5"
        >
          <li>
            <a
              href="/about"
            >
              About Zopa
            </a>
          </li>
          <li>
            <a
              href="/about/our-story"
            >
              Our story
            </a>
          </li>
          <li>
            <a
              href="/about/board"
            >
              Our board
            </a>
          </li>
          <li>
            <a
              href="/about/leadership"
            >
              Our leadership team
            </a>
          </li>
          <li>
            <a
              href="/about/awards"
            >
              Awards
            </a>
          </li>
          <li>
            <a
              href="/about/careers"
            >
              Careers
            </a>
          </li>
          <li>
            <a
              href="/about/press"
            >
              Press office
            </a>
          </li>
        </ul>
      </li>
      <li
        class="c6"
      >
        <a
          href="/contact"
        >
          Support
        </a>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="sc-gzVnrw cSqmIF"
    >
      <a
        aria-expanded="false"
        aria-haspopup="true"
        class="sc-htpNat Umdkr sc-ifAKCX bIyiel"
        color="#3B46C4"
        data-automation="ZA.navbar-item"
        href="#"
        tabindex="0"
      >
        <span
          class="sc-EHOje ccecCV"
        >
          About
        </span>
        <span
          class="sc-bZQynM DMXmU"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down fa-w-14 "
            color="#818F9B"
            data-icon="chevron-down"
            data-prefix="fas"
            focusable="false"
            height="12px"
            role="img"
            viewBox="0 0 448 512"
            width="12px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
              fill="currentColor"
            />
          </svg>
        </span>
      </a>
      <ul
        aria-label="About"
        class="sc-bdVaJa eQtwvl"
      >
        <li>
          <a
            href="/about"
          >
            About Zopa
          </a>
        </li>
        <li>
          <a
            href="/about/our-story"
          >
            Our story
          </a>
        </li>
        <li>
          <a
            href="/about/board"
          >
            Our board
          </a>
        </li>
        <li>
          <a
            href="/about/leadership"
          >
            Our leadership team
          </a>
        </li>
        <li>
          <a
            href="/about/awards"
          >
            Awards
          </a>
        </li>
        <li>
          <a
            href="/about/careers"
          >
            Careers
          </a>
        </li>
        <li>
          <a
            href="/about/press"
          >
            Press office
          </a>
        </li>
      </ul>
    </li>
    <li
      class="sc-htoDjs kzVlcq"
    >
      <a
        href="/contact"
      >
        Support
      </a>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
